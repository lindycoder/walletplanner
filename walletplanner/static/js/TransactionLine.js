import React from 'react';
import PropTypes from 'prop-types';
import CurrencyInput from 'react-currency-input';
import DatePicker from 'react-datepicker';


class TransactionLine extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            transaction: this.props.transaction
        }
    }

    handleAmountChange(event, maskedvalue, floatvalue) {
        const transaction = this.state.transaction;
        transaction.amount = Math.trunc(floatvalue * 100);

        this._update(transaction);
    }
    handleDateChange(date) {
        const transaction = this.state.transaction;
        transaction.date = date;

        this._update(transaction);
    }

    _update(transaction) {
        this.setState({transaction: transaction});

        if (typeof this.props.onChange === 'function') {
            this.props.onChange(this.state.transaction);
        }
    }

    render() {
        return (
            <div className="form-row">
              <div className="form-group col-sm-11">
               <DatePicker
                   selected={this.state.transaction.date}
                   onChange={this.handleDateChange.bind(this)}
                   className="transaction-line-date"
               />
               <CurrencyInput
                   value={this.format(this.state.transaction.amount)}
                   onChangeEvent={this.handleAmountChange.bind(this)}
                   className="transaction-line-amount"
               />
              </div>
            </div>
        );
    }

    format(amount) {
        return amount / 100;
    }
};


TransactionLine.propTypes = {
  transaction: PropTypes.any.isRequired,
  onChange: PropTypes.func
};

export default TransactionLine;