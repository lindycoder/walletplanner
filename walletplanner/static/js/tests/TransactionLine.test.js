import React from 'react';
import { shallow } from 'enzyme';
import TransactionLine from '../TransactionLine';

describe('TransactionLine', () => {
    let props;
    let wrapper;

    beforeEach(() => {
        props = {
            transaction: {
               "amount": 10,
               "date": new Date("2019-01-02T03:04:05.678"),
               "description": "desc1",
               "category": "cat1"
            },
            onChange: jest.fn()
        };
        wrapper = shallow(<TransactionLine {...props} />);
    });

    it('should render an initial value', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should invoke onChange when the input text changes', () => {
        wrapper.find('CurrencyInput').first().simulate('changeEvent', null, "100.00", 100.0);
        expect(props.onChange).toHaveBeenCalledWith(props.transaction);
        expect(props.transaction.amount).toBe(10000);
    });

    it('should invoke onChange when the date picker changes', () => {
        wrapper.find('DatePicker').first().simulate('change', new Date("2019-01-03T03:04:05.678"));
        expect(props.onChange).toHaveBeenCalledWith(props.transaction);
        expect(props.transaction.date).toEqual(new Date("2019-01-03T03:04:05.678"));
    });
});

describe('TransactionLine without OnChange', () => {
    let props;
    let wrapper;

    beforeEach(() => {
        props = {
            transaction: {
               "amount": 10,
               "description": "desc1",
               "category": "cat1"
            }
        };
        wrapper = shallow(<TransactionLine {...props} />);
    });

    it('should not fail when a change occurs without an onchange prop', () => {
        wrapper.find('CurrencyInput').first().simulate('changeEvent');
    });
});
