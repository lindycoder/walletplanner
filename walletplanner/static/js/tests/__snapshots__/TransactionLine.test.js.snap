// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionLine should render an initial value 1`] = `
<div
  className="form-row"
>
  <div
    className="form-group col-sm-11"
  >
    <DatePicker
      allowSameDay={false}
      className="transaction-line-date"
      dateFormat="MM/dd/yyyy"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      inlineFocusSelectedMonth={false}
      monthsShown={1}
      nextMonthButtonLabel="Next month"
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      readOnly={false}
      renderDayContents={[Function]}
      selected={2019-01-02T03:04:05.678Z}
      shouldCloseOnSelect={true}
      showMonthYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      withPortal={false}
    />
    <CurrencyInput
      allowNegative={false}
      autoFocus={false}
      className="transaction-line-amount"
      decimalSeparator="."
      inputType="text"
      onChange={[Function]}
      onChangeEvent={[Function]}
      precision="2"
      prefix=""
      selectAllOnFocus={false}
      suffix=""
      thousandSeparator=","
      value={0.1}
    />
  </div>
</div>
`;
